### PR TITLE
Fine grained notification for RecyclerViewAdapter

### PR DESCRIPTION
**Pull request by beeender**

### CODE REVIEW

File: adapters/src/main/java/io/realm/RealmBaseAdapter.java [hunk 1]
```diff
@@ -87,10 +87,12 @@ public void onChange(BaseRealm results) {
     private void addListener(@NonNull OrderedRealmCollection<T> data) {
         if (data instanceof RealmResults) {
```
Comment (cmelchior) [inline adapters/src/main/java/io/realm/RealmBaseAdapter.java]: check for instance of RealmCollectionObservable instead? and collapse the two if sections?

Comment (cmelchior) [inline adapters/src/main/java/io/realm/RealmBaseAdapter.java]: Ignore this after removing this interface

**Comment by cmelchior** [comment]
Currently crashes with:
```
02-27 09:44:35.634  2781  2794 E AndroidRuntime: FATAL EXCEPTION: Instr: android.support.test.runner.AndroidJUnitRunner
02-27 09:44:35.634  2781  2794 E AndroidRuntime: Process: io.realm.android.tests.test, PID: 2781
02-27 09:44:35.634  2781  2794 E AndroidRuntime: java.lang.IncompatibleClassChangeError: io.realm.AndroidNotifier
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at dalvik.system.DexFile.defineClassNative(Native Method)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at dalvik.system.DexFile.defineClass(DexFile.java:226)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at dalvik.system.DexFile.loadClassBinaryName(DexFile.java:219)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at dalvik.system.DexPathList.findClass(DexPathList.java:338)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:54)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at java.lang.Class.classForName(Native Method)
02-27 09:44:35.634  2781  2794 E AndroidRuntime: 	at java.lang.Class.forName(Class.java:324)
02
```

When running `gw clean :tests:connectedCheck`. Not entirely sure why.

**Comment by zaki50** [comment]
@cmelchior I could reproduce this with `3.0.0-SNAPSHOT` from jfrog but worked fine with my local build.

Something wrong in SNAPSHOT build?

**Comment by cmelchior** [comment]
Perhaps. I had to upload the SNAPSHOT from my local machine because CI didn't do it. Can you try running `ojoUpload` ?

File: adapters/src/main/java/io/realm/RealmRecyclerViewAdapter.java [hunk 2]
```diff
@@ -85,12 +79,7 @@ public RealmRecyclerViewAdapter(@Nullable OrderedRealmCollection<T> data, boolea
         // Right now don't use generics, since we need maintain two different
         // types of listeners until RealmList is properly supported.
         // See https://github.com/realm/realm-java/issues/989
```
Comment (nhachicha) [inline adapters/src/main/java/io/realm/RealmRecyclerViewAdapter.java]: is the above assumption still valid? https://github.com/realm/realm-java/issues/989  is closed now?

File: build.gradle [hunk 3]
```diff
@@ -22,11 +25,19 @@ buildscript {
     }
 }
 
+// Don't cache SNAPSHOT (changing) dependencies.
```
Comment (nhachicha) [inline build.gradle]: remember to remove when releasing

File: example/src/main/java/io/realm/examples/adapters/model/Counter.java [hunk 4]
```diff
@@ -15,19 +15,30 @@
...
 
+import java.util.concurrent.atomic.AtomicInteger;
+
 import io.realm.RealmObject;
 
-public class TimeStamp extends RealmObject {
+public class Counter extends RealmObject {
+    public static final String FIELD_COUNT = "count";
+
+    private static AtomicInteger integerCounter = new AtomicInteger(0);
```
Comment (nhachicha) [inline example/src/main/java/io/realm/examples/adapters/model/Counter.java]: why static? can't we use a member counter with `@Ignore` ?

File: example/src/main/java/io/realm/examples/adapters/model/Counter.java [hunk 5]
```diff
@@ -15,19 +15,30 @@
...
 
-    public String getTimeStamp() {
-        return timeStamp;
+    public String getCountString() {
+        return Integer.toString(count);
     }
 
-    public void setTimeStamp(String timeStamp) {
-        this.timeStamp = timeStamp;
+    public void setAndIncrease() {
```
Comment (nhachicha) [inline example/src/main/java/io/realm/examples/adapters/model/Counter.java]: I think `increment` is probably a more reflective name, as we're not setting any value (method doesn't take args)

File: example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java [hunk 6]
```diff
@@ -66,32 +68,81 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
     }
 
-    public void deleteItem(TimeStamp item) {
-        final String id = item.getTimeStamp();
+    // Randomly duplicate/delete/create some objects in the Realm.
+    private void randomEditItem() {
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                Random rand = new Random();
```
Comment (nhachicha) [inline example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java]: I think `Random` is a bit expensive to init. better create it as a member I guess.

File: example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java [hunk 7]
```diff
@@ -66,32 +68,81 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
+                RealmResults<Counter> results = realm.where(Counter.class).findAllSorted(Counter.FIELD_COUNT);
+
+                // Duplicate some existing entries.
+                int countToDup = results.size() > 3 ? 3 : results.size();
```
Comment (nhachicha) [inline example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java]: `countToDup` max is 3 (just double checking)

File: adapters/src/main/java/io/realm/RealmBaseAdapter.java [hunk 8]
```diff
@@ -34,47 +32,17 @@
...
-    }
+    private final RealmChangeListener<Object> listener;
 
     public RealmBaseAdapter(@Nullable OrderedRealmCollection<T> data) {
         if (data != null && !data.isManaged())
             throw new IllegalStateException("Only use this adapter with managed list, " +
                     "for un-managed lists you can just use the BaseAdapter");
         this.adapterData = data;
-        this.listener = new RealmChangeListener<BaseRealm>() {
+        this.listener = new RealmChangeListener<Object>() {
```
Comment (zaki50) [inline adapters/src/main/java/io/realm/RealmBaseAdapter.java]: Agree. I think it's OK to use raw type (`RealmChangeListener`). 
`RealmRecyclerViewAdapter` also is using raw type.

File: example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java [hunk 9]
```diff
@@ -66,32 +68,81 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
+                for (int i = 0; i < countToDup; i++) {
+                    int id = results.get(rand.nextInt((results.size()))).getCount();
```
Comment (zaki50) [inline example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java]: Why `id`?

This value is shared among multiple items.
It does not seem `id`.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,14 @@
+## 2.0.0 (YYYY-MM-DD)
```
Comment (kneth) [inline CHANGELOG.md]: Sounds like a good idea.

File: example/src/main/java/io/realm/examples/adapters/model/Counter.java (hunk 4 above)
Comment (cmelchior) [inline example/src/main/java/io/realm/examples/adapters/model/Counter.java]: I believe this is in order to have a global counter while the app is running so the list will have ever increasing numbers.

File: example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java (hunk 7 above)
Comment (cmelchior) [inline example/src/main/java/io/realm/examples/adapters/ui/recyclerview/RecyclerViewExampleActivity.java]: Yes, refactored to named variable for clarity

**Comment by cmelchior** [comment]
@nhachicha @zaki50 PR updated

**Pull request merged**